### DISCUSSION TITLE
Not able to access dotenv variables in the Strapi Admin front-end for different env's

### DISCUSSION THREAD

**Discussion by vinod77** [strapi/strapi#5599] [Legacy Topics]
**Describe the topic**
How can i access `dotenv` variables for `UAT` & `PROD`environments  in Strapi Admin front-end ? (I have overridden some front-end Admin code)
Currently in the Admin front-end I am always getting it as `process.env.NODE_ENV`: `production` in all env’s. But it should be different for each environment: `DEV`,  `UAT`,  & `PROD`

**Expected Behavior**
Should be able to access dotenv variables in all different environments & `process.env.NODE_ENV` should return respective `dotenv` variable

On DEV:  `process.env.NODE_ENV`: `dev`
On UAT:  `process.env.NODE_ENV`: `uat`
On PROD:  `process.env.NODE_ENV`: `production`
 
**Your suggestions for this topic**
Need to get access to `dotenv` variables for all respective environments (DEV, UAT, PROD).

- [ ] I have created an RFC on the [Strapi RFC Repo](https://github.com/strapi/rfcs)
- [ ] I have checked for existing RFCs before creating this discussion topic

**System**
- Strapi version: `3.0.0-beta.19.3`
- Node.js version: `v10.15.3`
- NPM version: `6.4.1`
- Database: `Mongo`
- Operating system: `MAC`

**Comment by alexandrebodin** [comment]
Hi That's not how the client side works :) The env variables are created at build time in the js bundle not at runtime.

**Comment by vinod77** [comment]
@alexandrebodin Thanks for your response. 
I know env variables are created at build time. But i should be able to access dotenv variables at build time as well.

**At Build Time** i want to get:
On DEV: process.env.NODE_ENV: dev
On UAT: process.env.NODE_ENV: uat
On PROD: process.env.NODE_ENV: production

But currently on all the environments i am getting `process.env.NODE_ENV` as  `production`. I want to get it as said above at build time.

**Comment by alexandrebodin** [comment]
This is not something we will be able to do. THe NODE_ENV=production is required to build the admin panel correctly. It doesn't recognize other env variables.

**Comment by vinod77** [comment]
@alexandrebodin Could you please let me know if there is any other way/workaround for this? 
Right now i have manually added our UAT `dotenv` variable in this file: `strapi-admin/admin/src/config.js`: https://github.com/strapi/strapi/blob/master/packages/strapi-admin/admin/src/config.js  as we are only on UAT now, it's working as expected on UAT.
But I need to add PROD `dotenv` variable as well before we go to PROD, not knowing how i can do that as i need to switch these UAT/PROD dotenv variables on UAT/PROD environments resp..

**Comment by soupette** [comment]
Hi @vinod77 now to we allow to customise the [webpack configuration](https://strapi.io/documentation/3.0.0-beta.x/admin-panel/custom-webpack-config.html) you can add env variable in the administration panel using webpack maybe it can help you achieve what you need?

You can do something as follows:

```
module.exports = {
  webpack: (config, webpack) => {
    const definePlugin = new webpack.DefinePlugin({
      MY_VAR: JSON.stringify('my var')
    });

    config.plugins.push(definePlugin);

    return config;
  }
}
```

I hope it helps

**Reply by divyansh0777** [reply]
Hello @soupette can you please let us know, how can we access this webpack under the Admin folder.

**Reply by soupette** [reply]
Hello here's how to do it:

1. First create a project

```bash
npx create-strapi-app my-app --quickstart --no-run
```

2. Create a `.env` file and add your environment variable

```bash
touch my-app/.env
```
**Path: `my-app/.env`**
```
MY_VAR=my_var
```

3. Extend the [webpack config](https://strapi.io/documentation/v3.x/admin-panel/custom-webpack-config.html)

```bash
mkdir  my-app/admin
touch my-app/admin/admin.config.js
```

**Path: `my-app/admin/admin.config.js`**

```js
module.exports = {
  webpack: (config, webpack) => {
    // Add your variable using the DefinePlugin
    config.plugins.push(
      new webpack.DefinePlugin({
        MY_VAR: JSON.stringify(process.env.MY_VAR),
      })
    );

    return config;
  },
};

```

4. Use your variable in the admin

In the extended files of the admin panel
```js
console.log(MY_VAR); // => 'my_var'
```

**Reply by divyansh0777** [reply]
@soupette @derrickmehaffy Please let me know, for which version this method supports?

I am currently using this one. Is it applicable for the below-listed version of strapi!
Node -> v13.7.0
Strapi -> v3.0.0-beta.18.4
OS -> Ubuntu 16.04

**Comment by derrickmehaffy** [comment]
Locking as migrated to: https://forum.strapi.io/t/not-able-to-access-dotenv-variables-in-the-strapi-admin-front-end-for-different-envs-5599/1751
